### PR TITLE
fix(datepicker): added valid date check

### PR DESCRIPTION
**Pull request by j-mcgregor**
### Proposed Changes

Initial bug: user could enter an invalid date range
Extra bug: While manually typing an invalid date, the `setState` triggered by `componentDidMount` would trigger an infinte loop as parent components tried to validate it and then update the component

How to test:

Datepicker:

- The bug applies to the datepicker dropdown, but worth testing in every example.
- I made a new example in the demo to mimic the bug in Log Browser (see screenshot)
- For any datepicker with `selectionRules`, if you try and manually enter a date that's outside of what's specified in its rules, then it will stay selected, use the last valid date as display (but not set it), and show a red `invalid` border

image

https://user-images.githubusercontent.com/22773767/139930393-12f55da7-2d0d-499f-b9fb-8f1da5897999.mov

-   [ ] The proposed changes are covered by unit tests
-   [ ] The potential breaking changes are clearly identified
-   [ ] [README.md](https://github.com/coveo/react-vapor/blob/master/README.md) is adjusted to reflect the proposed changes (if relevant)

### CODE REVIEW

**Review by coveobot — COMMENTED** [review]
Stale review, a new push occured on the branch. Check for the new status :arrow_down:

File: packages/react-vapor/src/components/datePicker/DatePicker.tsx [hunk 1]
```diff
@@ -61,8 +63,21 @@ export class DatePicker extends React.PureComponent<IDatePickerProps, {isSelecte
...
+
+            this.props.selectionRules?.forEach((rule: ICalendarSelectionRule) => {
+                if (
+                    (rule.isFor === CalendarSelectionRuleType.lower && this.props.isSelecting === DateLimits.lower) ||
+                    (rule.isFor === CalendarSelectionRuleType.upper && this.props.isSelecting === DateLimits.upper)
+                ) {
+                    isDatePermitted = rule.test(date);
+                    this.setState({isDatePermitted});
+                }
+            });
```
Comment (toofff) [inline packages/react-vapor/src/components/datePicker/DatePicker.tsx:78]: Use `filter` please

```suggestion
            this.props.selectionRules?
                .filter((rule: ICalendarSelectionRule) => (rule.isFor === CalendarSelectionRuleType.lower && this.props.isSelecting === DateLimits.lower) || (rule.isFor === CalendarSelectionRuleType.upper && this.props.isSelecting === DateLimits.upper))
                .forEach((rule: ICalendarSelectionRule) => {
                    isDatePermitted = rule.test(date);
                    this.setState({isDatePermitted});
                });
```

**Review by toofff — APPROVED** [review]

File: packages/demo/src/components/examples/DatePickerExamplesCommon.ts [hunk 2]
```diff
@@ -154,6 +154,21 @@ export const CALENDAR_SELECTION_RULES: ICalendarSelectionRule[] = [
     },
 ];
 
+export const CALENDAR_ADVANCED_SELECTION_RULES: ICalendarSelectionRule[] = [
+    {
+        test: (date: Date) => date <= new Date() && date >= moment().subtract(2, 'weeks').toDate(), // is the date within the last two weeks
```
Comment (aiheon) [inline packages/demo/src/components/examples/DatePickerExamplesCommon.ts:159]: should those rules have the range allowed as parameter?
I feel like we would need to add more rules for use case where the range is 3 days, 1 day, etc :thinking:

**Comment by aiheon** [comment]
image
This demo is broken also for the range :eyes: I can pick a bigger range from the calender directly even :sus:

File: packages/demo/src/components/examples/DatePickerExamplesCommon.ts (hunk 2 above)
Comment (aiheon) [reply packages/demo/src/components/examples/DatePickerExamplesCommon.ts:159]: Also, I think we should have a description popping at the same time that the red border if we are going down that route :thinking: 
Right now, it turns red but we don't know why (well we do but our users probably won't :see_no_evil: )

Comment (aiheon) [reply packages/demo/src/components/examples/DatePickerExamplesCommon.ts:159]: Maybe mimic this pattern instead? :eyes: 
image

Comment (j-mcgregor) [reply packages/demo/src/components/examples/DatePickerExamplesCommon.ts:159]: > should those rules have the range allowed as parameter? I feel like we would need to add more rules for use case where the range is 3 days, 1 day, etc 🤔

That was my first attempt but using range triggers the infinite loop

> Also, I think we should have a description popping at the same time that the red border if we are going down that route 🤔
Right now, it turns red but we don't know why (well we do but our users probably won't 🙈 )

I thought that too; I didn't attempt it yet as I wasn't sure what I have so far was the right way to go, and adding a message involves a fair bit more complexity

File: packages/react-vapor/src/components/datePicker/DatePicker.tsx [hunk 3]
```diff
@@ -61,8 +63,23 @@ export class DatePicker extends React.PureComponent<IDatePickerProps, {isSelecte
     private handleChangeDate = () => {
         if (this.dateInput.value !== '') {
             const date: Date = this.getDateFromString(this.dateInput.value);
-
-            if (date.getDate()) {
+            // setting isDatePermitted as let and in component state, since setState is async and triggers rerender of component
+            // and we need the value immediately to decide whether to to call props.onBlur
```
Comment (gdostie) [inline packages/react-vapor/src/components/datePicker/DatePicker.tsx:67]:
```suggestion
```

**Review by gdostie — COMMENTED** [review]
Add a unit test to make sure this edge case is covered and you have my approval

**Review by gdostie — APPROVED** [review]

**Review by GermainBergeron — APPROVED** [review]
